proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+will also help users to gain more control over accounts owned by them.
+
+Note that this proposal does not aim to increase gas fees for all transactions
+or add any kind of congestion control mechanism.
+
+## Alternatives Considered
+
+* Having a fixed write lock fee and a read lock fee.
+
+Pros: Simpler to implement, Simple to calculate fees.
+
+Cons: Will increase fees for everyone, dapp cannot decide to rebate fees, (or
+not all existing apps have to develop a rebate system). Fees cannot be decided
+by the dapp developer. Penalizes everyone for a few bad actors.
+
+* Passing fees for each account by instruction and validating them
+  during the execution inside user application code
+
+Pros: High efficiency, minimal performance impact, more generic, no need to
+change the account structure.
+
+Cons: Cannot prevent denial of service attacks, or account blocking
+attacks that do not call into the respective program
+
+## New Terminology
+
+Program Rebatable Account Write Fees or PRAW Fees: Fees that are decided by the
+dapp developer will be charged if a user wants to use the dapp. They will be
+applied even if the transaction fails contrary to lamport transfers. The program
+can decide to rebate these fees back to the user if certain conditions decided
+by dapp developers are met.
+
+### Other terminology
+
+Account `Authority`: The true authority for the account. Let's take an example
+of associated token account. A associated token account is a `PDA` derived from
+associated token program, where the owner is token program. But the program
+internally saves who has the authority over the account. Only token program can
+change the data of a token account. For operations like withdrawing tokens, the
+authority of the associated token account has to be the signer for the
+transaction containing transfer instruction. To receive tokens from another
+account, the authority's signature is not required.
+
+`Other Fees`: Fees other than PRAW fees `base fees + prioritization fees`.
+
+## Detailed Design
+
+PRAW fees enable dapp developer to decide who to rebate and who to penalize
+through runtime features. In particular, programs needs to be able to penalize,
+even if the application is not CPI'd to.
+
+The checks on the PRAW fees will be taken care of by Solana runtime. Total PRAW
+fees committed will be included in the transaction through a special parameter
+called `SetProgramRebatableAccountWriteFees` so that it is easier to calculate
+the total amount of fees that will be paid by the transaction. This parameter
+will be part of a new `TransactionHeaderProgram` where all the transaction
+parameters will be set through a TVL. In case of nonpayment or partial payment,
+the program is never executed. This instruction cannot be CPI'ed into.
+
+Addition of new syscalls to get, set and rebate PRAW fees during program
+runtime. Application developers will be able to rebate these fees through the
```

Comment:
(i noticed another clarification needs):

Addition of new syscalls to get, set and rebate PRAW fees during program
runtime.

=>

This proposal introduces new syscalls to individually get, set and rebate PRAW fees on addresses owned by a executing program. PRAW fees of accounts will be persisted into account database for the purpose of fee calculation of subsequent transactions. Also, the bank hash calculation is changed to incorporate it. Albeit storage increase for all accounts, rent won't be adjusted to retain compatibility.

Reply:
Will try to replace deprecated fields in the account structure typically rent epoch. If not possible adding new fields in accounts db will require a proposal of its own. Anyways this proposal aims to get consensus over PRAW fees and the UX of PRAW fees.